File: opensource/wallet_auth_flutter/lib/src/siwa/siwa_message.dart
import '../caip122/caip122_message.dart';

/// SIWA / EIP-4361 human-readable sign-in message (@avmkit/siwa compatible).
class SiwaMessage {
  const SiwaMessage({
    required this.domain,
    required this.accountAddress,
    required this.uri,
    required this.chainId,
    required this.nonce,
    this.statement,
    this.version = '1',
    this.issuedAt,
    this.expirationTime,
    this.notBefore,
    this.requestId,
    this.resources = const [],
    this.scheme,
  });

  factory SiwaMessage.fromCaip122(Caip122Message caip122, {required int walletConnectChainId}) {
    return SiwaMessage(
      domain: caip122.domain,
      accountAddress: caip122.accountAddress,
      uri: caip122.uri,
      chainId: walletConnectChainId,
      nonce: caip122.nonce,
      statement: caip122.statement,
      version: caip122.version,
      issuedAt: caip122.issuedAt,
      expirationTime: caip122.expirationTime,
      notBefore: caip122.notBefore,
      requestId: caip122.requestId,
      resources: caip122.resources,
    );
  }

  final String domain;
  final String accountAddress;
  final String uri;
  final int chainId;
  final String nonce;
  final String? statement;
  final String version;
  final String? issuedAt;
  final String? expirationTime;
  final String? notBefore;
  final String? requestId;
  final List<String> resources;
  final String? scheme;

  String prepareMessage() {
    if (statement != null && statement!.contains('\n')) {
      throw ArgumentError('statement must not contain newlines');
    }

    final headerPrefix = scheme != null ? '$scheme://$domain' : domain;
    final header = '$headerPrefix wants you to sign in with your Algorand account:';
    var prefix = '$header\n$accountAddress';

    if (statement != null && statement!.isNotEmpty) {
      prefix = '$prefix\n\n$statement\n';
    }

    final suffixParts = <String>[
      'URI: $uri',
      'Version: $version',
      'Chain ID: $chainId',
      'Nonce: $nonce',
    ];

    if (issuedAt != null) suffixParts.add('Issued At: $issuedAt');
    if (expirationTime != null) suffixParts.add('Expiration Time: $expirationTime');
    if (notBefore != null) suffixParts.add('Not Before: $notBefore');
    if (requestId != null) suffixParts.add('Request ID: $requestId');
    if (resources.isNotEmpty) {
      suffixParts.add('Resources:\n${resources.map((r) => '- $r').join('\n')}');
    }

    return '$prefix\n${suffixParts.join('\n')}';
  }
}

File: backend/app/modules/auth/utils/siwa_message.py
from __future__ import annotations

from app.modules.auth.utils.caip122 import Caip122Message


def prepare_siwa_message(
    *,
    domain: str,
    account_address: str,
    uri: str,
    chain_id: int,
    nonce: str,
    statement: str | None = None,
    version: str = "1",
    issued_at: str | None = None,
    expiration_time: str | None = None,
    not_before: str | None = None,
    request_id: str | None = None,
    resources: list[str] | None = None,
    scheme: str | None = None,
) -> str:
    """EIP-4361 / SIWA human-readable message (compatible with @avmkit/siwa)."""
    if statement and "\n" in statement:
        msg = "statement must not contain newlines"
        raise ValueError(msg)

    header_prefix = f"{scheme}://{domain}" if scheme else domain
    header = f"{header_prefix} wants you to sign in with your Algorand account:"
    prefix = f"{header}\n{account_address}"

    if statement:
        prefix = f"{prefix}\n\n{statement}\n"

    suffix_parts = [
        f"URI: {uri}",
        f"Version: {version}",
        f"Chain ID: {chain_id}",
        f"Nonce: {nonce}",
    ]

    if issued_at:
        suffix_parts.append(f"Issued At: {issued_at}")
    if expiration_time:
        suffix_parts.append(f"Expiration Time: {expiration_time}")
    if not_before:
        suffix_parts.append(f"Not Before: {not_before}")
    if request_id:
        suffix_parts.append(f"Request ID: {request_id}")
    if resources:
        resource_lines = "\n".join(f"- {r}" for r in resources)
        suffix_parts.append(f"Resources:\n{resource_lines}")

    return f"{prefix}\n{chr(10).join(suffix_parts)}"


def prepare_siwa_from_caip122(caip122: Caip122Message, *, wallet_connect_chain_id: int) -> str:
    """Build SIWA display string from CAIP-122 fields."""
    return prepare_siwa_message(
        domain=caip122.domain,
        account_address=caip122.account_address,
        uri=caip122.uri,
        chain_id=wallet_connect_chain_id,
        nonce=caip122.nonce,
        statement=caip122.statement,
        version=caip122.version,
        issued_at=caip122.issued_at,
        expiration_time=caip122.expiration_time,
        not_before=caip122.not_before,
        request_id=caip122.request_id,
        resources=caip122.resources or None,
    )
